tests/performance-tests/src/tests/public-api/data-set-query/queryGenerators.ts

```diff
@@ -0,0 +1,248 @@
+    };
+  }
+
+  private generateDataSetQueryNode({
+    dataSetMeta,
+    currentDepth,
+  }: {
+    dataSetMeta: DataSetMeta;
+    currentDepth: number;
+  }): DataSetQueryNode {
+    const allowedNodeTypes: DataSetQueryNodeType[] =
+      currentDepth === this.maxDepth
+        ? ['criteria', 'not']
+        : [...datSetQueryNodeTypes];
+    const nodeType = pickRandom(allowedNodeTypes);
+
+    switch (nodeType) {
+      case 'criteria':
+      case 'not': {
+        return this.generateDataSetCriteriaNode(dataSetMeta);
+      }
+      default: {
+        const childNodeCount = Math.max(
+          2,
+          Math.ceil(this.maxBranching * Math.random()),
+        );
+        return {
+          [nodeType]: Array.from({ length: childNodeCount }, _ =>
+            this.generateDataSetQueryNode({
+              dataSetMeta,
+              currentDepth: currentDepth + 1,
+            }),
+          ),
+        };
+      }
+    }
+  }
+
+  private generateIdListClause(
+    idList: string[],
+    range: Range,
+  ): DataSetQueryIdPredicate {
+    const operator = pickRandom(this.idOperators);
+    switch (operator) {
+      case 'in':
+      case 'notIn': {
+        return {
+          [operator]: pickRandomItems(
+            idList,
+            Math.min(range.random(), this.maxArrayItems),
+          ),
+        };
+      }
+      case 'eq':
+      case 'notEq': {
+        return {
+          [operator]: pickRandom(idList),
+        };
+      }
+      default: {
+        throw new Error(`Unsupported id operator ${operator}`);
+      }
+    }
+  }
+
+  private generateComparableListClause<T>(
+    comparableList: T[],
+    range: Range,
+  ): DataSetQueryComparablePredicate<T> {
+    const operator = pickRandom(this.comparableOperators);
+    switch (operator) {
+      case 'in':
+      case 'notIn': {
+        return {
+          [operator]: pickRandomItems(
+            comparableList,
+            Math.min(range.random(), this.maxArrayItems),
+          ),
+        };
+      }
+      case 'eq':
+      case 'notEq': {
+        return {
+          [operator]: pickRandom(comparableList),
+        };
+      }
+      case 'gt':
+      case 'gte': {
+        return {
+          [operator]: pickRandom(comparableList.slice(-1)),
+        };
+      }
+      case 'lt':
+      case 'lte': {
+        return {
+          [operator]: pickRandom(comparableList.slice(1)),
+        };
+      }
+      default: {
+        throw new Error(`Unsupported comparable operator ${operator}`);
+      }
+    }
+  }
+
+  private generateDataSetCriteriaNode(
+    dataSetMeta: DataSetMeta,
+  ): DataSetQueryCriteria {
+    const filterRange = new Range(0.1, 0.5);
+    const filterItemRange = new Range(0.1, 0.5);
+    const locationsRange = new Range(0.1, 0.5);
+    const timePeriodsRange = new Range(0.1, 0.5);
+    const geographicLevelRange = new Range(0.1, 0.5);
+
+    const flatLocations = Object.values(dataSetMeta.locations)
+      .flat()
+      .map(l => l.id);
```

Comment:
`flatMap`?

Reply:
Good call :)

Reply:
Ahh it's not a candidate for flatMap here - the initial flat() gives us a list of Locations from a list of lists of Locations (from Object.values()), and then we map() over them.